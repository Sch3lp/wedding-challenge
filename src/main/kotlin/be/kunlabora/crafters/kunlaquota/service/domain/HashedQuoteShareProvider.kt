package be.kunlabora.crafters.kunlaquota.service.domain

class HashedQuoteShareProvider: QuoteShareProvider {
    private val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    override fun invoke(quoteId: QuoteId): QuoteShare =
        quoteId.value.take(8)
            .hexToNumber().toLong()
            .encodeToBase26()
            .asQuoteShare()

    private fun String.asQuoteShare() = QuoteShare(this)

    @OptIn(ExperimentalStdlibApi::class)
    private fun String.hexToNumber(): String = hexToUInt(HexFormat.Default).toString()

    private fun Long.encodeToBase26(): String {
        val base = alphabet.length
        var num = this
        val result = StringBuilder()

        return if (num == 0L) {
            alphabet[0].toString()
        } else {
            while (num > 0) {
                val remainder = (num % base).toInt()
                result.append(alphabet[remainder])
                num /= base
            }

            result.reverse().toString()
        }
    }
}