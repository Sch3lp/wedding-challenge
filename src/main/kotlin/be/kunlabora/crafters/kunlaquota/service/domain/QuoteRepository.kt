package be.kunlabora.crafters.kunlaquota.service.domain

import be.kunlabora.crafters.kunlaquota.AddFailure
import be.kunlabora.crafters.kunlaquota.FetchQuotesFailed
import be.kunlabora.crafters.kunlaquota.service.Result

interface QuoteRepository {
    fun store(quote: Quote): Result<AddFailure, Quote>
    fun findAll(): Result<FetchQuotesFailed, List<Quote>>
}