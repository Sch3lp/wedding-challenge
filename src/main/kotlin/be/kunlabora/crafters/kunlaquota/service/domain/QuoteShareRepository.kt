package be.kunlabora.crafters.kunlaquota.service.domain

import be.kunlabora.crafters.kunlaquota.ShareFailure
import be.kunlabora.crafters.kunlaquota.service.Result

interface QuoteShareRepository {
    fun upsert(quoteId: QuoteId, quoteShareProvider: QuoteShareProvider): Result<ShareFailure, QuoteShare>
    fun findBy(quoteId: QuoteId): Result<ShareFailure, QuoteShare?>
    fun findBy(quoteShare: QuoteShare): Result<ShareFailure, QuoteId?>
}

typealias QuoteShareProvider = (QuoteId) -> QuoteShare
