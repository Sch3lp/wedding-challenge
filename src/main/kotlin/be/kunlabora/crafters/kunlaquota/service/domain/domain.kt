package be.kunlabora.crafters.kunlaquota.service.domain

import be.kunlabora.crafters.kunlaquota.service.EntityId
import java.time.LocalDateTime

typealias QuoteId = EntityId<Quote>

data class Quote(
    val id: QuoteId,
    val lines: List<Line> = emptyList(),
    val at: LocalDateTime = LocalDateTime.now(),
) {
    data class Line(
        val order: Int,
        val name: String,
        val text: String,
    )
}

@JvmInline value class QuoteShare(val value: String)