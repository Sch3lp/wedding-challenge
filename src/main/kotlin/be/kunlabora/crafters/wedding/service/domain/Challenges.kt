package be.kunlabora.crafters.wedding.service.domain

import be.kunlabora.crafters.wedding.ChallengeFailure
import be.kunlabora.crafters.wedding.service.Result

interface Challenges {
    fun store(challenge: Challenge): Challenge
    fun findAll(): List<Challenge>
    fun getById(challengeId: ChallengeId): Result<ChallengeFailure, Challenge>
}