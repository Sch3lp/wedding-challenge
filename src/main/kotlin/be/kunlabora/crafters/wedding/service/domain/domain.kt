package be.kunlabora.crafters.wedding.service.domain

import be.kunlabora.crafters.wedding.service.EntityId
import java.time.LocalDateTime

typealias ChallengeId = EntityId<Challenge>
typealias AssigneeId = EntityId<Assignee>

data class Challenge(
    val id: ChallengeId = ChallengeId.new(),
    val description: String,
    val assignee: AssigneeId,
    val completed: Boolean = false,
    val completedOn: LocalDateTime? = null,
) {
    fun markAsCompleted(): Challenge =
        copy(completed = true, completedOn = LocalDateTime.now())
}

data class Assignee(
    val name: String,
    val id: AssigneeId = AssigneeId.new(),
)