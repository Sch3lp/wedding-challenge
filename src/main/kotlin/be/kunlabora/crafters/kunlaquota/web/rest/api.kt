package be.kunlabora.crafters.kunlaquota.web.rest

import be.kunlabora.crafters.kunlaquota.AddFailure
import be.kunlabora.crafters.kunlaquota.QuoteAlreadyExists
import be.kunlabora.crafters.kunlaquota.QuoteIsInvalid
import be.kunlabora.crafters.kunlaquota.ShareQuoteFailed
import be.kunlabora.crafters.kunlaquota.service.*
import be.kunlabora.crafters.kunlaquota.service.domain.QuoteShare
import org.slf4j.LoggerFactory
import org.springframework.http.MediaType
import org.springframework.web.servlet.function.*
import java.net.URI

fun apiRoutes(quotes: IQuotes): RouterFunctionDsl.() -> Unit = {
    fun AddQuote.execute() = quotes.execute(this)
    fun ShareQuote.execute() = quotes.execute(this)

    "/quote".nest {
        GET(RequestPredicates.param("share") { true }) { request ->
            quotes.findByQuoteShare(QuoteShare(request.paramOrNull("share")!!))
                .map { foundQuotes ->
                    ServerResponse.ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(foundQuotes)
                }
                .recover { failure ->
                    ServerResponse.status(404).build().also { logger.error("$failure") }
                }.get()
        }

        GET("") {
            quotes.findAll()
                .map { foundQuotes ->
                    ServerResponse.ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(foundQuotes)
                }
                .recover { failure ->
                    ServerResponse.status(500).build().also { logger.error("$failure") }
                }.get()
        }

        POST("") { request ->
            request.body<AddQuote>()
                .execute()
                .map { quote ->
                    ServerResponse.created(quote.id.asUri(request)).build()
                }
                .handleAddQuoteFailure()
                .get()
        }

        POST("share") { request ->
            request.body<ShareQuote>()
                .execute()
                .map { quoteShare ->
                    ServerResponse.ok().body(quoteShare.asUri(request).toString())
                }
                .recover { failure ->
                    when (failure) {
                        is ShareQuoteFailed -> ServerResponse.status(500).build().also { logger.error(failure.message) }
                    }
                }
                .get()
        }
    }
}

private fun Result<AddFailure, ServerResponse>.handleAddQuoteFailure() =
    recover { failure ->
        when (failure) {
            is QuoteIsInvalid -> ServerResponse.status(400).body(AddQuoteError(failure.message))
            is QuoteAlreadyExists -> ServerResponse.status(500)
                .body(AddQuoteError("Something fishy occurred on our side. Maybe try again?"))
        }.also { logger.error(failure.message) }
    }


private val logger = LoggerFactory.getLogger("KunlaQuotaLogger")

private fun <E> EntityId<E>.asUri(request: ServerRequest): URI =
    request.uriBuilder().path("/{id}").build(this.value)

private fun QuoteShare.asUri(request: ServerRequest): URI =
    request.uriBuilder().replacePath("/api/quote").queryParam("share",this.value).build()