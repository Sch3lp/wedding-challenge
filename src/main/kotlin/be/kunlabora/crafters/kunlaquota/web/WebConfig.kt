package be.kunlabora.crafters.kunlaquota.web

import be.kunlabora.crafters.kunlaquota.service.IQuotes
import be.kunlabora.crafters.kunlaquota.web.rest.apiRoutes
import be.kunlabora.crafters.kunlaquota.web.ui.baseUiUrl
import be.kunlabora.crafters.kunlaquota.web.ui.uiRoutes
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer
import org.springframework.web.servlet.function.router


@Configuration
class WebConfig : WebMvcConfigurer {

    @Bean
    fun apiRoutesBean(quotes: IQuotes) = router {
        "/api".nest(apiRoutes(quotes))
    }

    @Bean
    fun uiRoutesBean(quotes: IQuotes) = router {
        baseUiUrl.nest(uiRoutes(quotes))
    }
}