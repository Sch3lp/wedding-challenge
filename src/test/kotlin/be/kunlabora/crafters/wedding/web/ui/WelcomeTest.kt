package be.kunlabora.crafters.wedding.web.ui

import be.kunlabora.crafters.wedding.service.domain.Assignee
import be.kunlabora.crafters.wedding.web.ui.test.verify
import be.kunlabora.crafters.wedding.web.welcome
import org.junit.jupiter.api.Test

class WelcomeTest {

    @Test
    fun `Welcome displays the assignee's name and a welcome message`() {
        verify { welcome(Assignee("snarf")) }
    }
}