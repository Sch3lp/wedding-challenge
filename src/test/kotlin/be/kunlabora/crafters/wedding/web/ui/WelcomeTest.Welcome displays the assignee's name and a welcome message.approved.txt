<p>Hello snarf. Enjoy the party!</p>
